src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -422,6 +424,17 @@ private static NamedTypeSymbol ValidateBuilderType(SyntheticBoundNodeFactory F,
                     !property.IsStatic &&
                     (property.ParameterCount == 0))
                 {
+                    HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                    Conversion conversion = F.Compilation.Conversions.ClassifyConversionFromType(property.Type, returnType, ref useSiteDiagnostics);
+                    if (conversion.Kind == ConversionKind.NoConversion)
```

Comment:
We could make this check even more restrictive and require that `property.Type == returnType`. Let me know.